test/test_pumactl.rb

```diff
@@ -106,50 +97,49 @@ def test_default_config_file_exist
 
   def test_control_no_token
     opts = [
-      "--config-file", "test/config/control_no_token.rb",
-      "start"
+      '--config-file', 'test/config/control_no_token.rb',
+      'start'
     ]
 
     control_cli = Puma::ControlCLI.new opts, @ready, @ready
-    assert_equal 'none', control_cli.instance_variable_get("@control_auth_token")
+    assert_equal 'none', control_cli.instance_variable_get('@control_auth_token')
   end
 
   def test_control_url_and_status
-    host = "127.0.0.1"
-    port = find_open_port
+    host = '127.0.0.1'
+    port = UniquePort.call
     url = "tcp://#{host}:#{port}/"
 
     opts = [
-      "--control-url", url,
-      "--control-token", "ctrl",
-      "--config-file", "test/config/app.rb",
+      '--control-url'  , url,
+      '--control-token', 'ctrl'
     ]
 
-    control_cli = Puma::ControlCLI.new (opts + ["start"]), @ready, @ready
+    control_cli = Puma::ControlCLI.new(opts + ['--config-file', 'test/config/app.rb', 'start'], @ready, @ready)
     t = Thread.new do
       control_cli.run
     end
 
-    wait_booted
+    assert_io 'Ctrl-C', io: @wait
 
     s = TCPSocket.new host, 9292
     s << "GET / HTTP/1.0\r\n\r\n"
     body = s.read
     assert_match "200 OK", body
     assert_match "embedded app", body
+    s.close
 
-    status_cmd = Puma::ControlCLI.new(opts + ["status"])
-    out, _ = capture_subprocess_io do
-      status_cmd.run
-    end
-    assert_match "Puma is started\n", out
+    Puma::ControlCLI.new(opts + ['status'], @ready, @ready).run
+    assert_io 'Puma is started', io: @wait
 
-    shutdown_cmd = Puma::ControlCLI.new(opts + ["halt"])
-    out, _ = capture_subprocess_io do
-      shutdown_cmd.run
-    end
-    assert_match "Command halt sent success\n", out
+    Puma::ControlCLI.new(opts + ['stop'], @ready, @ready).run
+    assert_io 'Goodbye', io: @wait
+
+    #Puma::ControlCLI.new(opts + ['halt'], @ready, @ready).run
+    #assert_io 'Stopping immediately!', io: @wait, debug: true
```

Comment:
Should this still be commented?

Reply:
IDK.  You may have seen errors where Events is trying to write but the pipe/IO is closed?  Those were part of what I was trying to clean up, so I may have switched to stop early in the work on it.

Also, IDK if we've got halt covered in all of the four classes that include it.  I'll look at tomorrow.

This PR was meant as a start to improving the test framework, both in terms of stability and ease of use.  I wanted it to be trivial to do tests for the same functionality, with one using a control socket and one using a signal.  I didn't want to mess with UNIX paths, remembering to close all the IO's, have leaky fd's, etc.  Also, no more naked read statements leading to lengthy 'test took too long' timeouts when a good method should limit them to 5 or 10 seconds, etc, etc..

I did have to use it in enough test files to see if it was functional, and it changed/expanded as I was developing it.  So, I'm sure this will change in the future, but I think it's a good first step.  

And, given how far I got into things with the Binder work and the systemd stuff, I feel pretty confident about the changes in the lib files.

Sorry for the extended post, I've wrestled with some messy things today.  There are a few small changes I'll push over the weekend.

Reply:
@nateberkopec

Hate to ask, can you try the above commented 'halt' lines with the two 'stop' lines commented?  Locally it freezes, probably why I commented it (a reminder for later).  I can push on my fork to Actions also...

Reply:
I rebased and deleted the comments.  We should see halt no being called in test coverage.